03_DeviceSelectionAndSharedSources/main.cpp

```diff
@@ -13,318 +13,422 @@ using namespace system;
+			options.preprocessorOptions.sourceIdentifier = source->getFilepathHint();
+			options.preprocessorOptions.logger = m_logger.get();
+			options.preprocessorOptions.includeFinder = compilerSet->getShaderCompiler(source->getContentType())->getDefaultIncludeFinder();
+
+			auto spirvUnspecialized = compilerSet->compileToSPIRV(source.get(), options);
+			const CSPIRVIntrospector::CStageIntrospectionData::SParams inspctParams = { .entryPoint = "main", .shader = spirvUnspecialized };
+
+			introspection = introspector.introspect(inspctParams);
+			if (!introspection)
 			{
-				IGPUQueue* const queue = getComputeQueue();
+				logFail("SPIR-V Introspection failed, probably the required SPIR-V compilation failed first!");
+				return std::pair(nullptr, nullptr);
+			}
 
-				smart_refctd_ptr<nbl::video::IGPUCommandBuffer> cmdbuf;
-				{
-					smart_refctd_ptr<nbl::video::IGPUCommandPool> cmdpool = m_device->createCommandPool(queue->getFamilyIndex(), IGPUCommandPool::ECF_TRANSIENT_BIT);
-					if (!m_device->createCommandBuffers(cmdpool.get(),IGPUCommandBuffer::EL_PRIMARY,1u,&cmdbuf))
-					{
-						logFail("Failed to create Command Buffers!\n");
-						return false;
-					}
-				
-					cmdbuf->begin(IGPUCommandBuffer::EU_ONE_TIME_SUBMIT_BIT);
-					cmdbuf->bindComputePipeline(pipeline.get());
-					cmdbuf->bindDescriptorSets(nbl::asset::EPBP_COMPUTE,pipeline->getLayout(),0,MaxDescriptorSets,&ds->get());
-					cmdbuf->dispatch(WorkgroupCount,1,1);
-					cmdbuf->end();
-				}
+			{
+				auto* srcContent = spirvUnspecialized->getContent();
 
-				// A sidenote: Waiting for the Device or Queue to become idle does not give the same memory visibility guarantees as a signalled fence or semaphore
-				smart_refctd_ptr<IGPUFence> done = m_device->createFence(IGPUFence::ECF_UNSIGNALED);
+				system::ISystem::future_t<core::smart_refctd_ptr<system::IFile>> future;
+				m_physicalDevice->getSystem()->createFile(future, system::path("../app_resources/compiled.spv"), system::IFileBase::ECF_WRITE);
+				if (auto file = future.acquire(); file && bool(*file))
 				{
-					IGPUQueue::SSubmitInfo submitInfo = {};
-					submitInfo.commandBufferCount = 1;
-					submitInfo.commandBuffers = &cmdbuf.get();
-
-					// To keep the sample renderdoc-able
-					queue->startCapture();
-					queue->submit(1u,&submitInfo,done.get());
-					queue->endCapture();
+					system::IFile::success_t succ;
+					(*file)->write(succ, srcContent->getPointer(), 0, srcContent->getSize());
+					succ.getBytesProcessed(true);
 				}
-				m_device->blockForFences(1,&done.get());
 			}
 
-			// a simple test to check we got the right thing back
-			auto outputData = mapBuffer(outputBuff,IDeviceMemoryAllocation::EMCAF_READ);
-			for (auto i=0; i<WorkgroupSize*WorkgroupCount; i++)
-			if (outputData[i]!=DWORDCount)
-				return logFail("DWORD at position %d doesn't match!\n",i);
-			m_device->unmapMemory(outputBuff->getBoundMemory());
-
-			return true;
```

Comment:
lets not loose this and actually do some dispatches with one of the compute shaders to do actual work like we did originally